kubernetes/apps_metrics_blackbox_test.go

```diff
@@ -0,0 +1,394 @@
+		// with the expected values
+		metrics := []*app.TimedNumberTuple{metric}
+		output := test.result.output
+		verifyMetrics(metrics, testCase, output, "network/rx_rate", t)
+
+		// Verify the remaining filters
+		verifySingleMetricFilters(testCase, output.filters, t)
+	}
+}
+
+func TestGetNetworkRecvRange(t *testing.T) {
+	testCases := metricRangeTestCases
+	test := &testHawkularGetter{}
+	config := &kubernetes.MetricsClientConfig{
+		MetricsURL:     "myMetricsServer",
+		BearerToken:    "token",
+		HawkularGetter: test,
+	}
+	client, err := kubernetes.NewMetricsClient(config)
+	require.NoError(t, err, "Failed to create metrics client")
+
+	for _, testCase := range testCases {
+		test.input = testCase
+		metrics, err := client.GetNetworkRecvMetricsRange(testCase.pods, testCase.namespace, testCase.startTime,
+			testCase.endTime, testCase.limit)
+		if !assert.NoError(t, err, "Getting network metrics failed") ||
+			!assert.NotNil(t, metrics, "Nil result from network metrics") {
+			continue
+		}
+
+		// Check that the result has the correct value and timestamp and that the Hawkular API was called
+		// with the expected values
+		output := test.result.output
+		verifyMetrics(metrics, testCase, output, "network/rx_rate", t)
+
+		// Verify the remaining filters
+		verifyMetricRangeFilters(testCase, output.filters, t)
+	}
+}
+
+func TestGetNetworkSent(t *testing.T) {
+	testCases := singleMetricTestCases
+	test := &testHawkularGetter{}
+	config := &kubernetes.MetricsClientConfig{
+		MetricsURL:     "myMetricsServer",
+		BearerToken:    "token",
+		HawkularGetter: test,
+	}
+	client, err := kubernetes.NewMetricsClient(config)
+	require.NoError(t, err, "Failed to create metrics client")
+
+	for _, testCase := range testCases {
+		test.input = testCase
+		metric, err := client.GetNetworkSentMetrics(testCase.pods, testCase.namespace, testCase.startTime)
+		if !assert.NoError(t, err, "Getting network metrics failed") ||
+			!assert.NotNil(t, metric, "Nil result from network metrics") {
+			continue
+		}
+
+		// Check that the result has the correct value and timestamp and that the Hawkular API was called
+		// with the expected values
+		metrics := []*app.TimedNumberTuple{metric}
+		output := test.result.output
+		verifyMetrics(metrics, testCase, output, "network/tx_rate", t)
+
+		// Verify the remaining filters
+		verifySingleMetricFilters(testCase, output.filters, t)
+	}
+}
+
+func TestGetNetworkSentRange(t *testing.T) {
+	testCases := metricRangeTestCases
+	test := &testHawkularGetter{}
+	config := &kubernetes.MetricsClientConfig{
+		MetricsURL:     "myMetricsServer",
+		BearerToken:    "token",
+		HawkularGetter: test,
+	}
+	client, err := kubernetes.NewMetricsClient(config)
+	require.NoError(t, err, "Failed to create metrics client")
+
+	for _, testCase := range testCases {
+		test.input = testCase
+		metrics, err := client.GetNetworkSentMetricsRange(testCase.pods, testCase.namespace, testCase.startTime, testCase.endTime, 0)
```

Comment:
Isn't this pretty much just testing that the TestGetter method will return what you asked it to? What about this when it comes through the API layer?

Reply:
It tests that metrics API calls for various inputs translate to the expected Hawkular API calls. It also checks that the data returned from Hawkular has the expected transformations applied to it.

The JSON API side of this PR just takes the return value from the metrics API and places it into a struct, but I still have more work to do writing a mock for the Kubernetes pod API before I can unit test this API method. Given the time constraints, I'd like to do this in a followup PR.